### DISCUSSION TITLE
How to save localstorage per story

### DISCUSSION THREAD

**Discussion by israelKusayev** [storybookjs/storybook#17766] [Help]
Is it possible to save something to `localStorage`/`sessionStorage` per story and/or per stories file?

Is there any addon for this, I can't find any

I would expect something like

```typescript
Default.parameters = {
	sessionStorage: {
		userId: '123'
	}
};
```

**Comment by iambumblehead** [comment]
_one, set localstorage using a decorator,_
```javascript
STORY.decorators = [ () => {
  window.localStorage.setItem( 'key', 'value' );
  return { template: '<story />' };
} ];
```
_or, two, set localstorage using a loader,_
```javascript
STORY.loaders = [ () => {
  window.localStorage.setItem( 'key', 'value' );
} ];
```

The "loader" can be defined as an async function while the "decorator" must be synchronous. Any localStorage definition will perist across stories, so it may be necessary to explicitly redefine localStorage for each story.

**Reply by boylec** [reply]
Easy way to reset local storage for every story

```typescript
// in preview.tsx

const localStorageResetDecorator: DecoratorFn = (Story) => {
  window.localStorage.clear();
  return <Story />;
};

export const decorators: DecoratorFn[] = [
   localStorageResetDecorator,
];

**Comment by artyrcheek** [comment]
Another option is to use a [loader](https://storybook.js.org/docs/writing-stories/loaders)
Here is a function I have for loading session storage into a story
```js 
function sessionStorageLoader(storage) {
  return async () => {
    Object.entries(storage).forEach(([key, value]) => {
      sessionStorage.setItem(key, JSON.stringify(value));
    });
  };
}
```
A pattern I find myself using is to clear the session storage keys I will be using in the default story, then override them for specific stories. e.g.
```js
export default {
  title: "Methods/Datepicker",
  component: Datepicker,
  loaders: [
    sessionStorageLoader({ startDate: null }),
  ],
}

export const WithDateSelected = {
  loaders: [
    sessionStorageLoader({ startDate: "2025-01-01 }),
  ],
};
```
